bauble_macro_util/src/lib.rs

```diff
@@ -0,0 +1,833 @@
+    enum FieldTy<'a> {
+        // The field may be deserialized from `bauble`, and must implement `FromBauble`
+        Val {
+            // An expression to generate this type. If `Some`, the field does not need to be
+            // specified in `bauble`.
+            default: Option<proc_macro2::TokenStream>,
+            // Whether the field is a `bauble` attribute
+            attribute: bool,
+            // Index for a tuple that holds the values of deserializable fields
+            index: Index,
+            // Type from which the field is deserialized
+            ty: &'a Type,
+        },
+        // The field is only generated from a default expression
+        AsDefault {
+            // An expression to generate this type
+            default: proc_macro2::TokenStream,
+            // Type from which the field is deserialized
+            ty: &'a Type,
+        },
+    }
+
+    // Information about a field collected from its attributes
+    struct FieldAttrs<'a> {
+        name: proc_macro2::TokenStream,
+        ty: FieldTy<'a>,
+    }
+
+    let mut val_count = 0;
+
+    // Parse the fields and attributes
+    let fields = match fields
+        .iter()
+        .enumerate()
+        .map(|(index, field)| -> Result<_, proc_macro2::TokenStream> {
+            let mut default = None;
+            let mut as_default = None;
+            let mut attribute = false;
+
+            for attr in &field.attrs {
+                if !attr.path().is_ident("bauble") {
+                    continue;
+                }
+
+                if let AttrStyle::Inner(_) = attr.style {
+                    Err(
+                        Error::new_spanned(attr, "inner attributes are not supported")
+                            .to_compile_error(),
+                    )?
+                }
+
+                attr.parse_nested_meta(|meta| {
+                    let Some(ident) = meta.path.get_ident() else {
+                        Err(meta.error("path must be an identifier"))?
+                    };
+
+                    match ident.to_string().as_str() {
+                        "default" => {
+                            if default.is_some() {
+                                Err(meta.error("duplicate `default` attribute"))?
+                            }
+
+                            if meta.input.parse::<Token![=]>().is_ok() {
+                                let expr = meta.input.parse::<Expr>()?;
+                                default = Some(quote! { #expr });
+                            } else {
+                                default = Some(quote! { ::std::default::Default::default() });
+                            }
+
+                            if !meta.input.is_empty() && !meta.input.peek(Token![,]) {
+                                Err(meta.error("unexpected token after default value"))?
+                            }
+
+                            Ok(())
+                        }
+                        "as_default" => {
```

Comment:
Is this just meant to be an alias for "default"?

Reply:
No; a `default` field may accept a value, but uses the default expression if not specified, whereas an `as_default` field may not accept a value and will always use the default expression. Notably, `as_default` fields do not need to implement `FromBauble`, which is useful for `PhantomData`.

If it helps, you can see these attributes used in practice [in `mgui`](https://github.com/Cakefish/dreamthorn/tree/feature/mgui_reactive_markup/mgui/mgui_backend/src)

Reply:
and in `mgui_backend`

Reply:
Well, the current code here for `as_default` seems to accept a value, so I'm guessing that's not intentional?

Reply:
It's intentional! What I mean by the value is a value specified in `bauble` data, and what I mean by default expression is an expression associated with the attribute